File: bindings/c/src/sign.c
/**
 * @file sign.c
 * @brief QUAC 100 SDK - Digital Signature Implementation
 * @copyright Copyright © 2025 Dyber, Inc. All Rights Reserved.
 */

#include "quac100/quac100.h"
#include "internal.h"

#include <stdlib.h>
#include <string.h>

/*============================================================================
 * Static Data
 *============================================================================*/

static const quac_sign_params_t g_sign_params[] = {
    /* ML-DSA (Dilithium) */
    {QUAC_SIGN_ML_DSA_44, 1312, 2560, 2420, 2, "ML-DSA-44"},
    {QUAC_SIGN_ML_DSA_65, 1952, 4032, 3309, 3, "ML-DSA-65"},
    {QUAC_SIGN_ML_DSA_87, 2592, 4896, 4627, 5, "ML-DSA-87"},

    /* SLH-DSA SHA2 */
    {QUAC_SIGN_SLH_DSA_SHA2_128S, 32, 64, 7856, 1, "SLH-DSA-SHA2-128s"},
    {QUAC_SIGN_SLH_DSA_SHA2_128F, 32, 64, 17088, 1, "SLH-DSA-SHA2-128f"},
    {QUAC_SIGN_SLH_DSA_SHA2_192S, 48, 96, 16224, 3, "SLH-DSA-SHA2-192s"},
    {QUAC_SIGN_SLH_DSA_SHA2_192F, 48, 96, 35664, 3, "SLH-DSA-SHA2-192f"},
    {QUAC_SIGN_SLH_DSA_SHA2_256S, 64, 128, 29792, 5, "SLH-DSA-SHA2-256s"},
    {QUAC_SIGN_SLH_DSA_SHA2_256F, 64, 128, 49856, 5, "SLH-DSA-SHA2-256f"},

    /* SLH-DSA SHAKE */
    {QUAC_SIGN_SLH_DSA_SHAKE_128S, 32, 64, 7856, 1, "SLH-DSA-SHAKE-128s"},
    {QUAC_SIGN_SLH_DSA_SHAKE_128F, 32, 64, 17088, 1, "SLH-DSA-SHAKE-128f"},
    {QUAC_SIGN_SLH_DSA_SHAKE_192S, 48, 96, 16224, 3, "SLH-DSA-SHAKE-192s"},
    {QUAC_SIGN_SLH_DSA_SHAKE_192F, 48, 96, 35664, 3, "SLH-DSA-SHAKE-192f"},
    {QUAC_SIGN_SLH_DSA_SHAKE_256S, 64, 128, 29792, 5, "SLH-DSA-SHAKE-256s"},
    {QUAC_SIGN_SLH_DSA_SHAKE_256F, 64, 128, 49856, 5, "SLH-DSA-SHAKE-256f"},
};

#define NUM_SIGN_PARAMS (sizeof(g_sign_params) / sizeof(g_sign_params[0]))

/*============================================================================
 * Parameter Functions
 *============================================================================*/

QUAC_API quac_status_t quac_sign_get_params(
    quac_sign_algorithm_t algorithm,
    quac_sign_params_t *params)
{
    QUAC_CHECK_PARAM(params != NULL);

    for (size_t i = 0; i < NUM_SIGN_PARAMS; i++)
    {
        if (g_sign_params[i].algorithm == algorithm)
        {
            *params = g_sign_params[i];
            return QUAC_SUCCESS;
        }
    }

    return QUAC_ERROR_NOT_SUPPORTED;
}

QUAC_API quac_status_t quac_sign_get_sizes(
    quac_sign_algorithm_t algorithm,
    size_t *public_key_len,
    size_t *secret_key_len,
    size_t *signature_len)
{
    quac_sign_params_t params;
    quac_status_t status = quac_sign_get_params(algorithm, &params);

    if (status != QUAC_SUCCESS)
    {
        return status;
    }

    if (public_key_len)
        *public_key_len = params.public_key_size;
    if (secret_key_len)
        *secret_key_len = params.secret_key_size;
    if (signature_len)
        *signature_len = params.signature_size;

    return QUAC_SUCCESS;
}

/*============================================================================
 * Signature Operations
 *============================================================================*/

QUAC_API quac_status_t quac_sign_keygen(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    uint8_t *public_key,
    size_t *public_key_len,
    uint8_t *secret_key,
    size_t *secret_key_len)
{
    QUAC_CHECK_INIT();
    QUAC_CHECK_DEVICE(device);
    QUAC_CHECK_PARAM(public_key != NULL);
    QUAC_CHECK_PARAM(public_key_len != NULL);
    QUAC_CHECK_PARAM(secret_key != NULL);
    QUAC_CHECK_PARAM(secret_key_len != NULL);

    quac_sign_params_t params;
    quac_status_t status = quac_sign_get_params(algorithm, &params);
    if (status != QUAC_SUCCESS)
    {
        return status;
    }

    if (*public_key_len < params.public_key_size ||
        *secret_key_len < params.secret_key_size)
    {
        *public_key_len = params.public_key_size;
        *secret_key_len = params.secret_key_size;
        return QUAC_ERROR_BUFFER_SMALL;
    }

    QUAC_LOG_DEBUG("Sign keygen: %s", params.name);

    quac_device_handle_impl *handle = (quac_device_handle_impl *)device;

    quac_device_lock_internal(handle);
    status = quac_core_sign_keygen(handle, algorithm,
                                   public_key, public_key_len,
                                   secret_key, secret_key_len);
    quac_device_unlock_internal(handle);

    if (status == QUAC_SUCCESS)
    {
        handle->total_ops++;
    }
    else
    {
        handle->total_errors++;
    }

    return status;
}

QUAC_API quac_status_t quac_sign(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *secret_key,
    size_t secret_key_len,
    const uint8_t *message,
    size_t message_len,
    uint8_t *signature,
    size_t *signature_len)
{
    QUAC_CHECK_INIT();
    QUAC_CHECK_DEVICE(device);
    QUAC_CHECK_PARAM(secret_key != NULL);
    QUAC_CHECK_PARAM(message != NULL || message_len == 0);
    QUAC_CHECK_PARAM(signature != NULL);
    QUAC_CHECK_PARAM(signature_len != NULL);

    quac_sign_params_t params;
    quac_status_t status = quac_sign_get_params(algorithm, &params);
    if (status != QUAC_SUCCESS)
    {
        return status;
    }

    if (secret_key_len != params.secret_key_size)
    {
        QUAC_LOG_ERROR("Invalid secret key size: %zu (expected %zu)",
                       secret_key_len, params.secret_key_size);
        return QUAC_ERROR_INVALID_KEY;
    }

    if (*signature_len < params.signature_size)
    {
        *signature_len = params.signature_size;
        return QUAC_ERROR_BUFFER_SMALL;
    }

    if (message_len > QUAC_MAX_MESSAGE_SIZE)
    {
        return QUAC_ERROR_INVALID_PARAM;
    }

    QUAC_LOG_DEBUG("Sign: %s, message_len=%zu", params.name, message_len);

    quac_device_handle_impl *handle = (quac_device_handle_impl *)device;

    quac_device_lock_internal(handle);
    status = quac_core_sign(handle, algorithm,
                            secret_key, secret_key_len,
                            message, message_len,
                            signature, signature_len);
    quac_device_unlock_internal(handle);

    if (status == QUAC_SUCCESS)
    {
        handle->total_ops++;
    }
    else
    {
        handle->total_errors++;
    }

    return status;
}

QUAC_API quac_status_t quac_verify(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *public_key,
    size_t public_key_len,
    const uint8_t *message,
    size_t message_len,
    const uint8_t *signature,
    size_t signature_len)
{
    QUAC_CHECK_INIT();
    QUAC_CHECK_DEVICE(device);
    QUAC_CHECK_PARAM(public_key != NULL);
    QUAC_CHECK_PARAM(message != NULL || message_len == 0);
    QUAC_CHECK_PARAM(signature != NULL);

    quac_sign_params_t params;
    quac_status_t status = quac_sign_get_params(algorithm, &params);
    if (status != QUAC_SUCCESS)
    {
        return status;
    }

    if (public_key_len != params.public_key_size)
    {
        QUAC_LOG_ERROR("Invalid public key size: %zu (expected %zu)",
                       public_key_len, params.public_key_size);
        return QUAC_ERROR_INVALID_KEY;
    }

    if (signature_len != params.signature_size)
    {
        QUAC_LOG_ERROR("Invalid signature size: %zu (expected %zu)",
                       signature_len, params.signature_size);
        return QUAC_ERROR_INVALID_PARAM;
    }

    QUAC_LOG_DEBUG("Verify: %s, message_len=%zu", params.name, message_len);

    quac_device_handle_impl *handle = (quac_device_handle_impl *)device;

    quac_device_lock_internal(handle);
    status = quac_core_verify(handle, algorithm,
                              public_key, public_key_len,
                              message, message_len,
                              signature, signature_len);
    quac_device_unlock_internal(handle);

    if (status == QUAC_SUCCESS || status == QUAC_ERROR_VERIFY_FAILED)
    {
        handle->total_ops++;
    }
    else
    {
        handle->total_errors++;
    }

    return status;
}

QUAC_API quac_status_t quac_sign_prehash(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *secret_key,
    size_t secret_key_len,
    const uint8_t *hash,
    size_t hash_len,
    quac_hash_algorithm_t hash_alg,
    uint8_t *signature,
    size_t *signature_len)
{
    (void)hash_alg;
    return quac_sign(device, algorithm, secret_key, secret_key_len,
                     hash, hash_len, signature, signature_len);
}

QUAC_API quac_status_t quac_verify_prehash(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *public_key,
    size_t public_key_len,
    const uint8_t *hash,
    size_t hash_len,
    quac_hash_algorithm_t hash_alg,
    const uint8_t *signature,
    size_t signature_len)
{
    (void)hash_alg;
    return quac_verify(device, algorithm, public_key, public_key_len,
                       hash, hash_len, signature, signature_len);
}

/*============================================================================
 * Batch Operations
 *============================================================================*/

QUAC_API quac_status_t quac_sign_keygen_batch(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    size_t count,
    uint8_t **public_keys,
    uint8_t **secret_keys,
    quac_status_t *status_codes)
{
    QUAC_CHECK_INIT();
    QUAC_CHECK_DEVICE(device);
    QUAC_CHECK_PARAM(count > 0);
    QUAC_CHECK_PARAM(public_keys != NULL);
    QUAC_CHECK_PARAM(secret_keys != NULL);
    QUAC_CHECK_PARAM(status_codes != NULL);

    quac_sign_params_t params;
    quac_status_t status = quac_sign_get_params(algorithm, &params);
    if (status != QUAC_SUCCESS)
    {
        return status;
    }

    quac_device_handle_impl *handle = (quac_device_handle_impl *)device;
    quac_device_lock_internal(handle);

    for (size_t i = 0; i < count; i++)
    {
        size_t pk_len = params.public_key_size;
        size_t sk_len = params.secret_key_size;

        status_codes[i] = quac_core_sign_keygen(handle, algorithm,
                                                public_keys[i], &pk_len,
                                                secret_keys[i], &sk_len);

        if (status_codes[i] == QUAC_SUCCESS)
        {
            handle->total_ops++;
        }
        else
        {
            handle->total_errors++;
        }
    }

    quac_device_unlock_internal(handle);
    return QUAC_SUCCESS;
}

QUAC_API quac_status_t quac_sign_batch(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *secret_key,
    size_t secret_key_len,
    size_t count,
    const uint8_t **messages,
    const size_t *message_lens,
    uint8_t **signatures,
    size_t *signature_lens,
    quac_status_t *status_codes)
{
    QUAC_CHECK_INIT();
    QUAC_CHECK_DEVICE(device);
    QUAC_CHECK_PARAM(secret_key != NULL);
    QUAC_CHECK_PARAM(count > 0);
    QUAC_CHECK_PARAM(messages != NULL);
    QUAC_CHECK_PARAM(message_lens != NULL);
    QUAC_CHECK_PARAM(signatures != NULL);
    QUAC_CHECK_PARAM(signature_lens != NULL);
    QUAC_CHECK_PARAM(status_codes != NULL);

    quac_sign_params_t params;
    quac_status_t status = quac_sign_get_params(algorithm, &params);
    if (status != QUAC_SUCCESS)
    {
        return status;
    }

    quac_device_handle_impl *handle = (quac_device_handle_impl *)device;
    quac_device_lock_internal(handle);

    for (size_t i = 0; i < count; i++)
    {
        signature_lens[i] = params.signature_size;

        status_codes[i] = quac_core_sign(handle, algorithm,
                                         secret_key, secret_key_len,
                                         messages[i], message_lens[i],
                                         signatures[i], &signature_lens[i]);

        if (status_codes[i] == QUAC_SUCCESS)
        {
            handle->total_ops++;
        }
        else
        {
            handle->total_errors++;
        }
    }

    quac_device_unlock_internal(handle);
    return QUAC_SUCCESS;
}

QUAC_API quac_status_t quac_verify_batch(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *public_key,
    size_t public_key_len,
    size_t count,
    const uint8_t **messages,
    const size_t *message_lens,
    const uint8_t **signatures,
    const size_t *signature_lens,
    int *results,
    quac_status_t *status_codes)
{
    QUAC_CHECK_INIT();
    QUAC_CHECK_DEVICE(device);
    QUAC_CHECK_PARAM(public_key != NULL);
    QUAC_CHECK_PARAM(count > 0);
    QUAC_CHECK_PARAM(messages != NULL);
    QUAC_CHECK_PARAM(message_lens != NULL);
    QUAC_CHECK_PARAM(signatures != NULL);
    QUAC_CHECK_PARAM(signature_lens != NULL);
    QUAC_CHECK_PARAM(results != NULL);
    QUAC_CHECK_PARAM(status_codes != NULL);

    quac_device_handle_impl *handle = (quac_device_handle_impl *)device;
    quac_device_lock_internal(handle);

    for (size_t i = 0; i < count; i++)
    {
        status_codes[i] = quac_core_verify(handle, algorithm,
                                           public_key, public_key_len,
                                           messages[i], message_lens[i],
                                           signatures[i], signature_lens[i]);

        results[i] = (status_codes[i] == QUAC_SUCCESS) ? 1 : 0;
        handle->total_ops++;
    }

    quac_device_unlock_internal(handle);
    return QUAC_SUCCESS;
}

/*============================================================================
 * Async Operations (Stub)
 *============================================================================*/

QUAC_API quac_status_t quac_sign_keygen_async(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    uint8_t *public_key,
    size_t *public_key_len,
    uint8_t *secret_key,
    size_t *secret_key_len,
    quac_async_callback_t callback,
    void *user_data,
    quac_async_t *async_handle)
{
    (void)callback;
    (void)user_data;
    (void)async_handle;
    return quac_sign_keygen(device, algorithm, public_key, public_key_len, secret_key, secret_key_len);
}

QUAC_API quac_status_t quac_sign_async(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *secret_key,
    size_t secret_key_len,
    const uint8_t *message,
    size_t message_len,
    uint8_t *signature,
    size_t *signature_len,
    quac_async_callback_t callback,
    void *user_data,
    quac_async_t *async_handle)
{
    (void)callback;
    (void)user_data;
    (void)async_handle;
    return quac_sign(device, algorithm, secret_key, secret_key_len,
                     message, message_len, signature, signature_len);
}

QUAC_API quac_status_t quac_verify_async(
    quac_device_t device,
    quac_sign_algorithm_t algorithm,
    const uint8_t *public_key,
    size_t public_key_len,
    const uint8_t *message,
    size_t message_len,
    const uint8_t *signature,
    size_t signature_len,
    quac_async_callback_t callback,
    void *user_data,
    quac_async_t *async_handle)
{
    (void)callback;
    (void)user_data;
    (void)async_handle;
    return quac_verify(device, algorithm, public_key, public_key_len,
                       message, message_len, signature, signature_len);
}

File: bindings/csharp/Sign.cs
// D:\quantacore-sdk\bindings\csharp\Sign.cs
// QUAC 100 SDK - Digital Signature Operations (ML-DSA, SLH-DSA)
// Copyright © 2025 Dyber, Inc. All Rights Reserved.

namespace Dyber.Quac100;

/// <summary>
/// ML-DSA and SLH-DSA Digital Signature operations
/// </summary>
public sealed class Signer
{
    private readonly Device _device;

    #region ML-DSA Parameters

    /// <summary>ML-DSA-44 parameters (NIST Level 2)</summary>
    public static readonly SignatureParameters MlDsa44 = new(
        SignatureAlgorithm.MlDsa44,
        publicKeySize: 1312,
        secretKeySize: 2560,
        signatureSize: 2420,
        securityLevel: 2,
        name: "ML-DSA-44"
    );

    /// <summary>ML-DSA-65 parameters (NIST Level 3, recommended)</summary>
    public static readonly SignatureParameters MlDsa65 = new(
        SignatureAlgorithm.MlDsa65,
        publicKeySize: 1952,
        secretKeySize: 4032,
        signatureSize: 3309,
        securityLevel: 3,
        name: "ML-DSA-65"
    );

    /// <summary>ML-DSA-87 parameters (NIST Level 5)</summary>
    public static readonly SignatureParameters MlDsa87 = new(
        SignatureAlgorithm.MlDsa87,
        publicKeySize: 2592,
        secretKeySize: 4896,
        signatureSize: 4627,
        securityLevel: 5,
        name: "ML-DSA-87"
    );

    #endregion

    #region SLH-DSA Parameters (SHA2)

    /// <summary>SLH-DSA-SHA2-128s parameters</summary>
    public static readonly SignatureParameters SlhDsaSha2_128s = new(
        SignatureAlgorithm.SlhDsaSha2_128s,
        publicKeySize: 32,
        secretKeySize: 64,
        signatureSize: 7856,
        securityLevel: 1,
        name: "SLH-DSA-SHA2-128s"
    );

    /// <summary>SLH-DSA-SHA2-128f parameters</summary>
    public static readonly SignatureParameters SlhDsaSha2_128f = new(
        SignatureAlgorithm.SlhDsaSha2_128f,
        publicKeySize: 32,
        secretKeySize: 64,
        signatureSize: 17088,
        securityLevel: 1,
        name: "SLH-DSA-SHA2-128f"
    );

    /// <summary>SLH-DSA-SHA2-192s parameters</summary>
    public static readonly SignatureParameters SlhDsaSha2_192s = new(
        SignatureAlgorithm.SlhDsaSha2_192s,
        publicKeySize: 48,
        secretKeySize: 96,
        signatureSize: 16224,
        securityLevel: 3,
        name: "SLH-DSA-SHA2-192s"
    );

    /// <summary>SLH-DSA-SHA2-192f parameters</summary>
    public static readonly SignatureParameters SlhDsaSha2_192f = new(
        SignatureAlgorithm.SlhDsaSha2_192f,
        publicKeySize: 48,
        secretKeySize: 96,
        signatureSize: 35664,
        securityLevel: 3,
        name: "SLH-DSA-SHA2-192f"
    );

    /// <summary>SLH-DSA-SHA2-256s parameters</summary>
    public static readonly SignatureParameters SlhDsaSha2_256s = new(
        SignatureAlgorithm.SlhDsaSha2_256s,
        publicKeySize: 64,
        secretKeySize: 128,
        signatureSize: 29792,
        securityLevel: 5,
        name: "SLH-DSA-SHA2-256s"
    );

    /// <summary>SLH-DSA-SHA2-256f parameters</summary>
    public static readonly SignatureParameters SlhDsaSha2_256f = new(
        SignatureAlgorithm.SlhDsaSha2_256f,
        publicKeySize: 64,
        secretKeySize: 128,
        signatureSize: 49856,
        securityLevel: 5,
        name: "SLH-DSA-SHA2-256f"
    );

    #endregion

    /// <summary>
    /// Create Signer operations context
    /// </summary>
    /// <param name="device">Device to use for operations</param>
    public Signer(Device device)
    {
        _device = device ?? throw new ArgumentNullException(nameof(device));
    }

    /// <summary>
    /// Get parameters for a signature algorithm
    /// </summary>
    public static SignatureParameters GetParameters(SignatureAlgorithm algorithm)
    {
        return algorithm switch
        {
            SignatureAlgorithm.MlDsa44 => MlDsa44,
            SignatureAlgorithm.MlDsa65 => MlDsa65,
            SignatureAlgorithm.MlDsa87 => MlDsa87,
            SignatureAlgorithm.SlhDsaSha2_128s => SlhDsaSha2_128s,
            SignatureAlgorithm.SlhDsaSha2_128f => SlhDsaSha2_128f,
            SignatureAlgorithm.SlhDsaSha2_192s => SlhDsaSha2_192s,
            SignatureAlgorithm.SlhDsaSha2_192f => SlhDsaSha2_192f,
            SignatureAlgorithm.SlhDsaSha2_256s => SlhDsaSha2_256s,
            SignatureAlgorithm.SlhDsaSha2_256f => SlhDsaSha2_256f,
            _ => throw new ArgumentException($"Unknown algorithm: {algorithm}", nameof(algorithm))
        };
    }

    /// <summary>
    /// Generate a new signature key pair
    /// </summary>
    /// <param name="algorithm">Signature algorithm variant</param>
    /// <returns>Generated key pair</returns>
    public KeyPair GenerateKeyPair(SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65)
    {
        var parameters = GetParameters(algorithm);

        var publicKey = new byte[parameters.PublicKeySize];
        var secretKey = new byte[parameters.SecretKeySize];
        int pkLen = publicKey.Length;
        int skLen = secretKey.Length;

        var status = (Quac100Status)NativeMethods.quac_sign_keygen(
            _device.Handle,
            (int)algorithm,
            publicKey, ref pkLen,
            secretKey, ref skLen
        );

        Quac100Exception.ThrowIfError(status, "Key generation failed");

        if (pkLen < publicKey.Length)
            Array.Resize(ref publicKey, pkLen);
        if (skLen < secretKey.Length)
            Array.Resize(ref secretKey, skLen);

        var keyPair = new KeyPair(publicKey, secretKey, algorithm)
        {
            Usage = KeyUsage.AllSign
        };

        return keyPair;
    }

    /// <summary>
    /// Generate key pair asynchronously
    /// </summary>
    public Task<KeyPair> GenerateKeyPairAsync(
        SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65,
        CancellationToken cancellationToken = default)
    {
        return Task.Run(() => GenerateKeyPair(algorithm), cancellationToken);
    }

    /// <summary>
    /// Sign a message
    /// </summary>
    /// <param name="secretKey">Signing key</param>
    /// <param name="message">Message to sign</param>
    /// <param name="algorithm">Signature algorithm</param>
    /// <returns>Signature bytes</returns>
    public byte[] Sign(ReadOnlySpan<byte> secretKey, ReadOnlySpan<byte> message, SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65)
    {
        var parameters = GetParameters(algorithm);

        if (secretKey.Length != parameters.SecretKeySize)
            throw new ArgumentException($"Invalid secret key size. Expected {parameters.SecretKeySize}, got {secretKey.Length}", nameof(secretKey));

        var signature = new byte[parameters.SignatureSize];
        int sigLen = signature.Length;

        var status = (Quac100Status)NativeMethods.quac_sign(
            _device.Handle,
            (int)algorithm,
            secretKey.ToArray(), secretKey.Length,
            message.ToArray(), message.Length,
            signature, ref sigLen
        );

        Quac100Exception.ThrowIfError(status, "Signing failed");

        if (sigLen < signature.Length)
            Array.Resize(ref signature, sigLen);

        return signature;
    }

    /// <summary>
    /// Sign a message using key pair
    /// </summary>
    public byte[] Sign(KeyPair keyPair, ReadOnlySpan<byte> message)
    {
        if (keyPair.Algorithm is not SignatureAlgorithm algorithm)
            throw new ArgumentException("Key pair is not a signature key pair", nameof(keyPair));

        return Sign(keyPair.SecretKey, message, algorithm);
    }

    /// <summary>
    /// Sign a string message (UTF-8 encoded)
    /// </summary>
    public byte[] Sign(KeyPair keyPair, string message)
    {
        return Sign(keyPair, System.Text.Encoding.UTF8.GetBytes(message));
    }

    /// <summary>
    /// Sign asynchronously
    /// </summary>
    public Task<byte[]> SignAsync(
        ReadOnlyMemory<byte> secretKey,
        ReadOnlyMemory<byte> message,
        SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65,
        CancellationToken cancellationToken = default)
    {
        return Task.Run(() => Sign(secretKey.Span, message.Span, algorithm), cancellationToken);
    }

    /// <summary>
    /// Verify a signature
    /// </summary>
    /// <param name="publicKey">Verification key</param>
    /// <param name="message">Original message</param>
    /// <param name="signature">Signature to verify</param>
    /// <param name="algorithm">Signature algorithm</param>
    /// <returns>True if signature is valid</returns>
    public bool Verify(ReadOnlySpan<byte> publicKey, ReadOnlySpan<byte> message, ReadOnlySpan<byte> signature, SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65)
    {
        var parameters = GetParameters(algorithm);

        if (publicKey.Length != parameters.PublicKeySize)
            throw new ArgumentException($"Invalid public key size. Expected {parameters.PublicKeySize}, got {publicKey.Length}", nameof(publicKey));

        var status = (Quac100Status)NativeMethods.quac_verify(
            _device.Handle,
            (int)algorithm,
            publicKey.ToArray(), publicKey.Length,
            message.ToArray(), message.Length,
            signature.ToArray(), signature.Length
        );

        if (status == Quac100Status.Success)
            return true;

        if (status == Quac100Status.VerifyFailed)
            return false;

        Quac100Exception.ThrowIfError(status, "Verification error");
        return false;
    }

    /// <summary>
    /// Verify a signature using key pair
    /// </summary>
    public bool Verify(KeyPair keyPair, ReadOnlySpan<byte> message, ReadOnlySpan<byte> signature)
    {
        if (keyPair.Algorithm is not SignatureAlgorithm algorithm)
            throw new ArgumentException("Key pair is not a signature key pair", nameof(keyPair));

        return Verify(keyPair.PublicKey, message, signature, algorithm);
    }

    /// <summary>
    /// Verify a signature (string message)
    /// </summary>
    public bool Verify(KeyPair keyPair, string message, ReadOnlySpan<byte> signature)
    {
        return Verify(keyPair, System.Text.Encoding.UTF8.GetBytes(message), signature);
    }

    /// <summary>
    /// Verify and throw if invalid
    /// </summary>
    public void VerifyOrThrow(ReadOnlySpan<byte> publicKey, ReadOnlySpan<byte> message, ReadOnlySpan<byte> signature, SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65)
    {
        if (!Verify(publicKey, message, signature, algorithm))
            throw new SignatureVerificationException("Signature verification failed");
    }

    /// <summary>
    /// Verify and throw if invalid (using key pair)
    /// </summary>
    public void VerifyOrThrow(KeyPair keyPair, ReadOnlySpan<byte> message, ReadOnlySpan<byte> signature)
    {
        if (!Verify(keyPair, message, signature))
            throw new SignatureVerificationException("Signature verification failed");
    }

    /// <summary>
    /// Verify asynchronously
    /// </summary>
    public Task<bool> VerifyAsync(
        ReadOnlyMemory<byte> publicKey,
        ReadOnlyMemory<byte> message,
        ReadOnlyMemory<byte> signature,
        SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65,
        CancellationToken cancellationToken = default)
    {
        return Task.Run(() => Verify(publicKey.Span, message.Span, signature.Span, algorithm), cancellationToken);
    }

    /// <summary>
    /// Perform a full sign/verify round-trip for testing
    /// </summary>
    public (KeyPair KeyPair, byte[] Signature, bool Verified) DemoSignVerify(
        string message = "Hello, Post-Quantum World!",
        SignatureAlgorithm algorithm = SignatureAlgorithm.MlDsa65)
    {
        var messageBytes = System.Text.Encoding.UTF8.GetBytes(message);

        // Generate key pair
        var keyPair = GenerateKeyPair(algorithm);

        // Sign message
        var signature = Sign(keyPair, messageBytes);

        // Verify signature
        var verified = Verify(keyPair, messageBytes, signature);

        return (keyPair, signature, verified);
    }

    /// <summary>
    /// Batch sign multiple messages
    /// </summary>
    public byte[][] SignBatch(KeyPair keyPair, ReadOnlySpan<byte>[] messages)
    {
        if (keyPair.Algorithm is not SignatureAlgorithm algorithm)
            throw new ArgumentException("Key pair is not a signature key pair", nameof(keyPair));

        var signatures = new byte[messages.Length][];

        for (int i = 0; i < messages.Length; i++)
        {
            signatures[i] = Sign(keyPair.SecretKey, messages[i], algorithm);
        }

        return signatures;
    }

    /// <summary>
    /// Batch verify multiple signatures
    /// </summary>
    public bool[] VerifyBatch(KeyPair keyPair, ReadOnlySpan<byte>[] messages, ReadOnlySpan<byte>[] signatures)
    {
        if (messages.Length != signatures.Length)
            throw new ArgumentException("Messages and signatures arrays must have the same length");

        if (keyPair.Algorithm is not SignatureAlgorithm algorithm)
            throw new ArgumentException("Key pair is not a signature key pair", nameof(keyPair));

        var results = new bool[messages.Length];

        for (int i = 0; i < messages.Length; i++)
        {
            results[i] = Verify(keyPair.PublicKey, messages[i], signatures[i], algorithm);
        }

        return results;
    }

    /// <summary>
    /// Parallel batch signing
    /// </summary>
    public async Task<byte[][]> SignBatchAsync(
        KeyPair keyPair,
        ReadOnlyMemory<byte>[] messages,
        int maxParallelism = 4,
        CancellationToken cancellationToken = default)
    {
        if (keyPair.Algorithm is not SignatureAlgorithm algorithm)
            throw new ArgumentException("Key pair is not a signature key pair", nameof(keyPair));

        var signatures = new byte[messages.Length][];
        var semaphore = new SemaphoreSlim(maxParallelism);

        var tasks = messages.Select(async (msg, index) =>
        {
            await semaphore.WaitAsync(cancellationToken);
            try
            {
                signatures[index] = await SignAsync(keyPair.SecretKey.ToArray(), msg, algorithm, cancellationToken);
            }
            finally
            {
                semaphore.Release();
            }
        });

        await Task.WhenAll(tasks);
        return signatures;
    }

    /// <summary>
    /// Detect algorithm from public key size
    /// </summary>
    public static SignatureAlgorithm? DetectAlgorithmFromPublicKey(int keySize)
    {
        return keySize switch
        {
            1312 => SignatureAlgorithm.MlDsa44,
            1952 => SignatureAlgorithm.MlDsa65,
            2592 => SignatureAlgorithm.MlDsa87,
            32 => SignatureAlgorithm.SlhDsaSha2_128s, // or 128f - need context
            48 => SignatureAlgorithm.SlhDsaSha2_192s, // or 192f
            64 => SignatureAlgorithm.SlhDsaSha2_256s, // or 256f
            _ => null
        };
    }

    /// <summary>
    /// Detect algorithm from secret key size
    /// </summary>
    public static SignatureAlgorithm? DetectAlgorithmFromSecretKey(int keySize)
    {
        return keySize switch
        {
            2560 => SignatureAlgorithm.MlDsa44,
            4032 => SignatureAlgorithm.MlDsa65,
            4896 => SignatureAlgorithm.MlDsa87,
            64 => SignatureAlgorithm.SlhDsaSha2_128s,
            96 => SignatureAlgorithm.SlhDsaSha2_192s,
            128 => SignatureAlgorithm.SlhDsaSha2_256s,
            _ => null
        };
    }
}

/// <summary>
/// Extension methods for signature operations
/// </summary>
public static class SignerExtensions
{
    /// <summary>
    /// Create Signer operations from device
    /// </summary>
    public static Signer GetSigner(this Device device) => new(device);

    /// <summary>
    /// Check if key pair is a signature key pair
    /// </summary>
    public static bool IsSignatureKeyPair(this KeyPair keyPair) => keyPair.Algorithm is SignatureAlgorithm;

    /// <summary>
    /// Get signature algorithm from key pair
    /// </summary>
    public static SignatureAlgorithm? GetSignatureAlgorithm(this KeyPair keyPair) => keyPair.Algorithm as SignatureAlgorithm?;
}